File: src/main/java/com/openlattice/chronicle/services/enrollment/EnrollmentService.java
package com.openlattice.chronicle.services.enrollment;

import com.google.common.base.Optional;
import com.google.common.collect.*;
import com.openlattice.ApiUtil;
import com.openlattice.chronicle.data.ChronicleCoreAppConfig;
import com.openlattice.chronicle.data.ChronicleDataCollectionAppConfig;
import com.openlattice.chronicle.data.ParticipationStatus;
import com.openlattice.chronicle.services.ApiCacheManager;
import com.openlattice.chronicle.services.ScheduledTasksManager;
import com.openlattice.chronicle.services.edm.EdmCacheManager;
import com.openlattice.chronicle.services.entitysets.EntitySetIdsManager;
import com.openlattice.chronicle.sources.AndroidDevice;
import com.openlattice.chronicle.sources.Datasource;
import com.openlattice.chronicle.sources.IOSDevice;
import com.openlattice.client.ApiClient;
import com.openlattice.data.*;
import com.openlattice.data.requests.NeighborEntityDetails;
import com.openlattice.search.SearchApi;
import com.openlattice.search.requests.EntityNeighborsFilter;
import org.apache.olingo.commons.api.edm.FullQualifiedName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.access.AccessDeniedException;

import java.util.*;
import java.util.concurrent.ExecutionException;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.openlattice.chronicle.constants.EdmConstants.MODEL_FQN;
import static com.openlattice.chronicle.constants.EdmConstants.NAME_FQN;
import static com.openlattice.chronicle.constants.EdmConstants.OL_ID_FQN;
import static com.openlattice.chronicle.constants.EdmConstants.STATUS_FQN;
import static com.openlattice.chronicle.constants.EdmConstants.STRING_ID_FQN;
import static com.openlattice.chronicle.constants.EdmConstants.VERSION_FQN;
import static com.openlattice.chronicle.util.ChronicleServerUtil.ORG_STUDY_PARTICIPANT;
import static com.openlattice.chronicle.util.ChronicleServerUtil.ORG_STUDY_PARTICIPANT_DATASOURCE;
import static com.openlattice.chronicle.util.ChronicleServerUtil.getFirstValueOrNull;
import static com.openlattice.chronicle.util.ChronicleServerUtil.getParticipantEntitySetName;

/**
 * @author alfoncenzioka &lt;alfonce@openlattice.com&gt;
 */
public class EnrollmentService implements EnrollmentManager {
    protected static final Logger logger = LoggerFactory.getLogger( EnrollmentService.class );

    private final ApiCacheManager       apiCacheManager;
    private final ScheduledTasksManager scheduledTasksManager;
    private final EdmCacheManager       edmCacheManager;
    private final EntitySetIdsManager   entitySetIdsManager;

    public EnrollmentService(
            ApiCacheManager apiCacheManager,
            EdmCacheManager edmCacheManager,
            EntitySetIdsManager entitySetIdsManager,
            ScheduledTasksManager scheduledTasksManager ) {

        this.edmCacheManager = edmCacheManager;
        this.scheduledTasksManager = scheduledTasksManager;
        this.apiCacheManager = apiCacheManager;
        this.entitySetIdsManager = entitySetIdsManager;
    }

    private UUID reserveDeviceEntityKeyId(
            UUID deviceEntitySetId,
            Map<UUID, Set<Object>> data,
            DataIntegrationApi dataIntegrationApi ) {

        EntityKey entityKey = new EntityKey(
                deviceEntitySetId,
                ApiUtil.generateDefaultEntityId(
                        ImmutableList.of( edmCacheManager.getPropertyTypeId( STRING_ID_FQN ) ),
                        data
                )
        );

        return dataIntegrationApi.getEntityKeyIds( ImmutableSet.of( entityKey ) ).iterator().next();
    }

    private void updateDeviceIdsCache( UUID organizationId, UUID deviceEKID, String datasourceId ) {
        if ( organizationId != null ) {
            scheduledTasksManager
                    .getDeviceIdsByOrg().computeIfAbsent( organizationId, key -> Maps.newHashMap() )
                    .put( datasourceId, deviceEKID );
            return;
        }

        scheduledTasksManager.getDeviceIdsByEKID().put( datasourceId, deviceEKID );
    }

    private Map<UUID, Set<Object>> createDeviceData( String datasourceId, Optional<Datasource> datasource ) {
        Map<UUID, Set<Object>> data = Maps.newHashMap();
        data.put( edmCacheManager.getPropertyTypeId( STRING_ID_FQN ), ImmutableSet.of( datasourceId ) );

        if ( datasource.isPresent() ) {
            if ( AndroidDevice.class.isAssignableFrom( datasource.get().getClass() ) ) {
                AndroidDevice device = (AndroidDevice) datasource.get();
                data.put( edmCacheManager.getPropertyTypeId( MODEL_FQN ), ImmutableSet.of( device.getModel() ) );
                data.put( edmCacheManager.getPropertyTypeId( VERSION_FQN ), ImmutableSet.of( device.getOsVersion() ) );
            }

            if ( IOSDevice.class.isAssignableFrom( datasource.get().getClass() ) ) {
                IOSDevice device = (IOSDevice) datasource.get();
                data.put( edmCacheManager.getPropertyTypeId( MODEL_FQN ),
                        ImmutableSet.of( device.getLocalizedModel() ) );
                data.put( edmCacheManager.getPropertyTypeId( VERSION_FQN ), ImmutableSet.of( device.getVersion() ) );
                data.put( edmCacheManager.getPropertyTypeId( NAME_FQN ), ImmutableSet.of( device.getName() ) );
            }
        }

        return data;
    }

    private UUID registerDatasourceHelper(
            UUID organizationId,
            UUID studyId,
            String participantId,
            String datasourceId,
            Optional<Datasource> datasource ) {

        try {
            ApiClient apiClient = apiCacheManager.prodApiClientCache.get( ApiClient.class );
            DataApi dataApi = apiClient.getDataApi();
            DataIntegrationApi dataIntegrationApi = apiClient.getDataIntegrationApi();

            // entity set ids
            ChronicleDataCollectionAppConfig dataCollectionAppConfig = entitySetIdsManager
                    .getChronicleDataCollectionAppConfig( organizationId );
            ChronicleCoreAppConfig coreAppConfig = entitySetIdsManager
                    .getChronicleAppConfig( organizationId, getParticipantEntitySetName( studyId ) );

            UUID usedByESID = dataCollectionAppConfig.getUsedByEntitySetId();
            UUID studyESID = coreAppConfig.getStudiesEntitySetId();
            UUID devicesESID = dataCollectionAppConfig.getDeviceEntitySetId();
            UUID participantsESID = coreAppConfig.getParticipantEntitySetId();

            // ensure study and participant exist
            UUID studyEKID = checkNotNull( getStudyEntityKeyId( organizationId, studyId ), "study must exist" );
            UUID participantEKID = checkNotNull( getParticipantEntityKeyId( organizationId, studyId, participantId ) );

            // device entity data

            Map<UUID, Set<Object>> deviceData = createDeviceData( datasourceId, datasource );

            UUID deviceEntityKeyId = reserveDeviceEntityKeyId( devicesESID, deviceData, dataIntegrationApi );
            if ( deviceEntityKeyId == null ) {
                logger.error(
                        "unable to reserve ekid for datasource" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                        organizationId,
                        studyId,
                        participantId,
                        datasourceId
                );
                return null;
            }
            dataApi.updateEntitiesInEntitySet( devicesESID,
                    ImmutableMap.of( deviceEntityKeyId, deviceData ),
                    UpdateType.Merge,
                    PropertyUpdateType.Versioned );

            updateDeviceIdsCache( organizationId, deviceEntityKeyId, datasourceId );

            EntityDataKey deviceEDK = new EntityDataKey( devicesESID, deviceEntityKeyId );
            EntityDataKey participantEDK = new EntityDataKey( participantsESID, participantEKID );
            EntityDataKey studyEDK = new EntityDataKey( studyESID, studyEKID );

            ListMultimap<UUID, DataEdge> associations = ArrayListMultimap.create();

            Map<UUID, Set<Object>> usedByEntity = ImmutableMap
                    .of( edmCacheManager.getPropertyTypeId( STRING_ID_FQN ), Sets.newHashSet( UUID.randomUUID() ) );

            associations.put( usedByESID, new DataEdge( deviceEDK, participantEDK, usedByEntity ) );
            associations.put( usedByESID, new DataEdge( deviceEDK, studyEDK, usedByEntity ) );

            dataApi.createAssociations( associations );

            logger.info(
                    "datasource registered" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                    organizationId,
                    studyId,
                    participantId,
                    datasourceId
            );
            return deviceEntityKeyId;
        } catch ( Exception exception ) {
            logger.error(
                    "unable to register datasource" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                    organizationId,
                    studyId,
                    participantId,
                    datasourceId
            );
            throw new RuntimeException( "unable to register datasource" );
        }
    }

    @Override
    public UUID registerDatasource(
            UUID organizationId,
            UUID studyId,
            String participantId,
            String datasourceId,
            Optional<Datasource> datasource ) {

        logger.info(
                "attempting to register data source" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                organizationId,
                studyId,
                participantId,
                datasourceId
        );

        final boolean isKnownParticipant = isKnownParticipant( organizationId, studyId, participantId );
        if ( !isKnownParticipant ) {
            logger.error(
                    "unknown participant, unable to register datasource" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                    organizationId,
                    studyId,
                    participantId,
                    datasourceId
            );
            throw new AccessDeniedException( "unknown participant, unable to register datasource" );
        }

        final UUID deviceEKID = getDeviceEntityKeyId( organizationId, studyId, participantId, datasourceId );
        if ( deviceEKID != null ) {
            logger.info(
                    "datasource is registered" + ORG_STUDY_PARTICIPANT_DATASOURCE,
                    organizationId,
                    studyId,
                    participantId,
                    datasourceId
            );
            return deviceEKID;
        }

        return registerDatasourceHelper( organizationId, studyId, participantId, datasourceId, datasource );
    }

    @Override
    public UUID getDeviceEntityKeyId( UUID organizationId, UUID studyId, String participantId, String datasourceId ) {
        if ( organizationId != null ) {
            return scheduledTasksManager
                    .getDeviceIdsByOrg()
                    .getOrDefault( organizationId, Map.of() )
                    .getOrDefault( datasourceId, null );
        }
        return scheduledTasksManager.getDeviceIdsByEKID().getOrDefault( datasourceId, null );
    }

    @Override
    public boolean isKnownDatasource( UUID organizationId, UUID studyId, String participantId, String datasourceId ) {
        return getDeviceEntityKeyId( organizationId, studyId, participantId, datasourceId ) != null;
    }

    @Override
    public boolean isKnownParticipant( UUID organizationId, UUID studyId, String participantId ) {
        return getParticipantEntityKeyId( organizationId, studyId, participantId ) != null;
    }

    @Override
    public Map<FullQualifiedName, Set<Object>> getParticipantEntity(
            UUID organizationId, UUID studyId, UUID participantEntityKeyId ) {
        try {
            ApiClient apiClient = apiCacheManager.prodApiClientCache.get( ApiClient.class );
            DataApi dataApi = apiClient.getDataApi();

            ChronicleCoreAppConfig coreAppConfig = entitySetIdsManager
                    .getChronicleAppConfig( organizationId, getParticipantEntitySetName( studyId ) );

            UUID entitySetId = coreAppConfig.getParticipantEntitySetId();
            if ( entitySetId == null ) {
                logger.error( "Unable to ge participant ESID: orgId = {}, studyId = {}, participantEKID = {}",
                        organizationId,
                        studyId,
                        participantEntityKeyId );
                return ImmutableMap.of();
            }
            Map<FullQualifiedName, Set<Object>> entity = dataApi.getEntity( entitySetId, participantEntityKeyId );
            if ( entity == null ) {
                logger.error(
                        "Unable to get participant entity: orgId = {}, studyId = {} participantEKID = {}, participantESID = {}",
                        organizationId,
                        studyId,
                        participantEntityKeyId,
                        entitySetId );
                return ImmutableMap.of();
            }
            return entity;

        } catch ( ExecutionException e ) {
            logger.error( "Unable to get participant entity: orgId = {}, studyId = {}, participantEKID = {} ",
                    organizationId,
                    studyId,
                    participantEntityKeyId,
                    e );
            return ImmutableMap.of();
        }
    }

    @Override
    public ParticipationStatus getParticipationStatus( UUID organizationId, UUID studyId, String participantId ) {

        ParticipationStatus status;

        logger.info( "getting participation status" + ORG_STUDY_PARTICIPANT, organizationId, studyId, participantId );

        try {
            ApiClient apiClient = apiCacheManager.prodApiClientCache.get( ApiClient.class );
            SearchApi searchApi = apiClient.getSearchApi();

            // entity set ids
            ChronicleCoreAppConfig coreAppConfig = entitySetIdsManager
                    .getChronicleAppConfig( organizationId, getParticipantEntitySetName( studyId ) );

            UUID studiesESID = coreAppConfig.getStudiesEntitySetId();
            UUID participatedInESID = coreAppConfig.getParticipatedInEntitySetId();
            UUID participantsESID = coreAppConfig.getParticipantEntitySetId();

            // participant must exist
            UUID participantEKID = checkNotNull(
                    getParticipantEntityKeyId( organizationId, studyId, participantId ),
                    "participant not found"
            );

            // filtered search on participants to get associated study entities
            Map<UUID, List<NeighborEntityDetails>> neighborResults = searchApi.executeFilteredEntityNeighborSearch(
                    participantsESID,
                    new EntityNeighborsFilter(
                            ImmutableSet.of( participantEKID ),
                            java.util.Optional.of( ImmutableSet.of() ),
                            java.util.Optional.of( ImmutableSet.of( studiesESID ) ),
                            java.util.Optional.of( ImmutableSet.of( participatedInESID ) )
                    )
            );

            status = neighborResults
                    .getOrDefault( participantEKID, List.of() )
                    .stream()
                    .filter( neighbor -> studyId.toString().equals(
                            getFirstValueOrNull( neighbor.getNeighborDetails().orElse( Map.of() ), STRING_ID_FQN )
                    ) )
                    .map( neighbor -> getFirstValueOrNull( neighbor.getAssociationDetails(), STATUS_FQN ) )
                    .filter( Objects::nonNull )
                    .map( ParticipationStatus::valueOf )
                    .findFirst()
                    .orElse( ParticipationStatus.UNKNOWN );

            logger.info(
                    "participation status = {}" + ORG_STUDY_PARTICIPANT,
                    status,
                    organizationId,
                    studyId,
                    participantId
            );

            return status;
        } catch ( Exception exception ) {
            logger.error(
                    "unable to get participation status, returning {}" + ORG_STUDY_PARTICIPANT,
                    ParticipationStatus.UNKNOWN,
                    organizationId,
                    studyId,
                    participantId,
                    exception
            );
            return ParticipationStatus.UNKNOWN;
        }
    }

    @Override
    public boolean isNotificationsEnabled( UUID organizationId, UUID studyId ) {
        logger.info( "Checking notifications enabled on studyId = {}, organization = {}", studyId, organizationId );

        try {
            ApiClient apiClient = apiCacheManager.prodApiClientCache.get( ApiClient.class );
            SearchApi searchApi = apiClient.getSearchApi();

            // get entity set ids
            ChronicleCoreAppConfig coreAppConfig = entitySetIdsManager.getChronicleAppConfig( organizationId );

            UUID studyESID = coreAppConfig.getStudiesEntitySetId();
            UUID notificationESID = coreAppConfig.getNotificationEntitySetId();
            UUID partOfESID = coreAppConfig.getPartOfEntitySetId();

            // ensure study exists
            UUID studyEKID = checkNotNull( getStudyEntityKeyId( organizationId, studyId ),
                    "study does not exist: orgId=%s, studyId=%s", organizationId, studyId );

            Map<UUID, List<NeighborEntityDetails>> neighbors = searchApi
                    .executeFilteredEntityNeighborSearch(
                            studyESID,
                            new EntityNeighborsFilter(
                                    ImmutableSet.of( studyEKID ),
                                    java.util.Optional.of( ImmutableSet.of( notificationESID ) ),
                                    java.util.Optional.of( ImmutableSet.of( studyESID ) ),
                                    java.util.Optional.of( ImmutableSet.of( partOfESID ) )
                            )
                    );

            /*
             * studies with notifications enabled have 'ol.id' property in the corresponding
             * associationDetails set to the value of the studyId
             * For each study, there is can only be at most 1 notification -> partof -> study association,
             * therefore it suffices to explore the first neighbor
             */
            return neighbors
                    .values()
                    .stream()
                    .flatMap( Collection::stream )
                    .map( NeighborEntityDetails::getAssociationDetails )
                    .anyMatch( neighbor -> studyId.toString().equals( getFirstValueOrNull( neighbor, OL_ID_FQN ) ) );

        } catch ( Exception e ) {
            String error =
                    "failed to get notification enabled status for study " + studyId + " in org " + organizationId;
            logger.error( error, e );
            throw new RuntimeException( error );
        }
    }

    @Override
    public UUID getParticipantEntityKeyId( UUID organizationId, UUID studyId, String participantId ) {
        if ( organizationId != null ) {
            Map<UUID, Map<String, UUID>> participants = scheduledTasksManager.getStudyParticipantsByOrg()
                    .getOrDefault( organizationId, Map.of() );

            return participants.getOrDefault( studyId, Map.of() ).getOrDefault( participantId, null );
        }

        return scheduledTasksManager.getStudyParticipants().getOrDefault( studyId, Map.of() )
                .getOrDefault( participantId, null );
    }

    @Override
    public Set<UUID> getStudyParticipants( UUID organizationId, UUID studyId ) {
        if ( organizationId != null ) {
            Map<UUID, Map<String, UUID>> participants = scheduledTasksManager.getStudyParticipantsByOrg()
                    .getOrDefault( organizationId, Map.of() );

            return Sets.newHashSet( participants.getOrDefault( studyId, Map.of() ).values() );
        }

        return Sets
                .newHashSet( scheduledTasksManager.getStudyParticipants().getOrDefault( studyId, Map.of() ).values() );
    }

    @Override
    public UUID getStudyEntityKeyId( UUID organizationId, UUID studyId ) {
        if ( organizationId != null ) {
            return scheduledTasksManager.getStudyEntityKeyIdsByOrg().getOrDefault( organizationId, Map.of() )
                    .getOrDefault( studyId, null );
        }
        return scheduledTasksManager.getStudyEKIDById().getOrDefault( studyId, null );
    }
}

File: src/main/kotlin/com/openlattice/chronicle/services/enrollment/EnrollmentService.kt
package com.openlattice.chronicle.services.enrollment

import com.geekbeast.controllers.exceptions.ResourceNotFoundException
import com.geekbeast.mappers.mappers.ObjectMappers
import com.geekbeast.postgres.PostgresArrays
import com.geekbeast.postgres.streams.BasePostgresIterable
import com.geekbeast.postgres.streams.PreparedStatementHolderSupplier
import com.openlattice.chronicle.data.ParticipationStatus
import com.openlattice.chronicle.ids.HazelcastIdGenerationService
import com.openlattice.chronicle.participants.Participant
import com.openlattice.chronicle.postgres.ResultSetAdapters
import com.openlattice.chronicle.services.candidates.CandidateManager
import com.openlattice.chronicle.sources.AndroidDevice
import com.openlattice.chronicle.sources.IOSDevice
import com.openlattice.chronicle.sources.SourceDevice
import com.openlattice.chronicle.sources.SourceDeviceType
import com.openlattice.chronicle.storage.ChroniclePostgresTables.Companion.DEVICES
import com.openlattice.chronicle.storage.ChroniclePostgresTables.Companion.STUDY_PARTICIPANTS
import com.openlattice.chronicle.storage.PostgresColumns.Companion.DEVICE_ID
import com.openlattice.chronicle.storage.PostgresColumns.Companion.DEVICE_TOKEN
import com.openlattice.chronicle.storage.PostgresColumns.Companion.DEVICE_TYPE
import com.openlattice.chronicle.storage.PostgresColumns.Companion.PARTICIPANT_ID
import com.openlattice.chronicle.storage.PostgresColumns.Companion.PARTICIPATION_STATUS
import com.openlattice.chronicle.storage.PostgresColumns.Companion.SOURCE_DEVICE_ID
import com.openlattice.chronicle.storage.PostgresColumns.Companion.STUDY_ID
import com.openlattice.chronicle.storage.StorageResolver
import com.openlattice.chronicle.util.ChronicleServerUtil
import org.slf4j.LoggerFactory
import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Service
import java.sql.Connection
import java.util.*
import javax.xml.transform.Source

/**
 * @author alfoncenzioka &lt;alfonce@openlattice.com&gt;
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
@Service
class EnrollmentService(
    private val storageResolver: StorageResolver,
    private val idGenerationService: HazelcastIdGenerationService,
    private val candidateManager: CandidateManager,
) : EnrollmentManager {

    companion object {
        private val logger = LoggerFactory.getLogger(EnrollmentService::class.java)
        private val DEVICES_COLS = DEVICES.columns.joinToString(",") { it.name }
        private val STUDY_PARTICIPANT_COLS = STUDY_PARTICIPANTS.columns.joinToString(",") { it.name }
        private val mapper = ObjectMappers.newJsonMapper()

        /**
         * 1. study id
         * 2. device id
         * 3. participant id
         * 4. device type
         * 5. source device id
         * 6. source device
         * 7. device token
         */
        private val INSERT_DEVICE = """
            INSERT INTO ${DEVICES.name} ($DEVICES_COLS) VALUES (?,?,?,?,?,?::jsonb,?) 
            ON CONFLICT (${STUDY_ID.name},${PARTICIPANT_ID.name}, ${SOURCE_DEVICE_ID.name}) DO UPDATE SET ${DEVICE_TOKEN.name} = EXCLUDED.${DEVICE_TOKEN.name}             
        """

        /**
         * 1. study id
         * 2. participant id
         * 3. source device id
         */
        private val GET_DEVICE_ID = """
            SELECT ${DEVICE_ID.name} FROM ${DEVICES.name} 
                WHERE ${STUDY_ID.name} = ? AND ${PARTICIPANT_ID.name} = ? AND ${SOURCE_DEVICE_ID.name} = ? 
        """.trimIndent()

        /**
         * Retrieves the distinct enrolled device types
         * 1. study id
         * 2. participant id
         */
        private val GET_DEVICE_TYPES = """
            SELECT DISTINCT ${STUDY_ID.name}, ${PARTICIPANT_ID.name}, array_agg(distinct ${DEVICE_TYPE.name}) as ${DEVICE_TYPE.name}
            FROM ${DEVICES.name}
            WHERE ${STUDY_ID.name} = ANY(?) AND ${PARTICIPANT_ID.name} = ANY(?)
            GROUP BY ${STUDY_ID.name}, ${PARTICIPANT_ID.name}
        """.trimIndent()

        /**
         * 1. study id
         * 2. participant id
         * 3. source device id
         */
        private val COUNT_DEVICE_ID = """
            SELECT count(*) FROM ${DEVICES.name} 
                WHERE ${STUDY_ID.name} = ? AND ${PARTICIPANT_ID.name} = ? AND ${SOURCE_DEVICE_ID.name} = ? 
        """.trimIndent()

        private val COUNT_STUDY_PARTICIPANTS = """
            SELECT count(*) FROM ${STUDY_PARTICIPANTS.name} WHERE ${STUDY_ID.name} = ? AND ${PARTICIPANT_ID.name} = ?
        """.trimIndent()

        /**
         * 1. study id
         * 2. participant id
         * 3. candidate id
         * 4. participation status
         */
        private val INSERT_PARTICIPANT = """
            INSERT INTO ${STUDY_PARTICIPANTS.name} ($STUDY_PARTICIPANT_COLS) VALUES (?,?,?,?)
        """.trimIndent()

        /**
         * 1. study id
         * 2. participant id
         */
        private val GET_PARTICIPANT = """
            SELECT * FROM ${STUDY_PARTICIPANTS.name} WHERE ${STUDY_ID.name} = ? AND ${PARTICIPANT_ID.name} = ?
        """.trimIndent()

        /**
         * 1. study id
         * 2. participant id
         */
        private val GET_PARTICIPATION_STATUS = """
            SELECT ${PARTICIPATION_STATUS.name} FROM ${STUDY_PARTICIPANTS.name} WHERE ${STUDY_ID.name} = ? AND ${PARTICIPANT_ID.name} = ?
        """.trimIndent()

    }

    override fun registerDevice(
        studyId: UUID,
        participantId: String,
        sourceDeviceId: String,
        sourceDevice: SourceDevice,
    ): UUID {
        logger.info(
            "attempting to register data source" + ChronicleServerUtil.STUDY_PARTICIPANT_DATASOURCE,
            studyId,
            participantId,
            sourceDeviceId
        )

        if (!isKnownParticipant(studyId, participantId)) {
            logger.error(
                "unknown participant, unable to register datasource" + ChronicleServerUtil.STUDY_PARTICIPANT_DATASOURCE,
                studyId,
                participantId,
                sourceDeviceId
            )
            throw AccessDeniedException("unknown participant, unable to register datasource")
        }


        return when (sourceDevice) {
            is AndroidDevice -> registerDeviceOrGetId(
                studyId,
                participantId,
                SourceDeviceType.Android,
                sourceDeviceId,
                sourceDevice,
                sourceDevice.fcmRegistrationToken
            )

            is IOSDevice -> registerDeviceOrGetId(
                studyId,
                participantId,
                SourceDeviceType.Ios,
                sourceDeviceId,
                sourceDevice,
                sourceDevice.apnDeviceToken
            )

            else -> throw UnsupportedOperationException("${sourceDevice.javaClass.name} is not a supported datasource.")
        }
    }

    private fun registerDeviceOrGetId(
        studyId: UUID,
        participantId: String,
        deviceType: SourceDeviceType,
        sourceDeviceId: String,
        sourceDevice: SourceDevice,
        deviceToken: String,
    ): UUID {
        val hds = storageResolver.getPlatformStorage()
        val deviceId = idGenerationService.getNextId()
        val insertCount = hds.connection.use { connection ->
            connection.prepareStatement(INSERT_DEVICE).use { ps ->
                ps.setObject(1, studyId)
                ps.setObject(2, deviceId)
                ps.setString(3, participantId)
                ps.setString(4, deviceType.name)
                ps.setString(5, sourceDeviceId)
                ps.setString(6, mapper.writeValueAsString(sourceDevice))
                ps.setString(7, deviceToken)
                ps.executeUpdate()
            }
        }

        if (insertCount > 0) {
            return deviceId
        }
        logger.info(
            "Datasource for ${ChronicleServerUtil.STUDY_PARTICIPANT_DATASOURCE} is already registered. Returning",
            studyId,
            participantId,
            sourceDeviceId
        )
        return getDeviceId(studyId, participantId, sourceDeviceId)
    }

    override fun getDeviceId(studyId: UUID, participantId: String, sourceDeviceId: String): UUID {
        val hds = storageResolver.getPlatformStorage()

        return hds.connection.use { connection ->
            connection.prepareStatement(GET_DEVICE_ID).use { ps ->
                ps.setObject(1, studyId)
                ps.setString(2, participantId)
                ps.setString(3, sourceDeviceId)
                ps.executeQuery().use { rs ->
                    if (rs.next()) {
                        ResultSetAdapters.deviceId(rs)
                    } else {
                        throw ResourceNotFoundException("Unable to find device for study=$studyId, participant=$participantId, sourceDeviceId=$sourceDeviceId")
                    }
                }
            }
        }
    }

    override fun registerParticipant(
        connection: Connection,
        studyId: UUID,
        participantId: String,
        candidateId: UUID,
        participationStatus: ParticipationStatus,
    ) {
        connection.prepareStatement(INSERT_PARTICIPANT).use { ps ->
            ps.setObject(1, studyId)
            ps.setString(2, participantId)
            ps.setObject(3, candidateId)
            ps.setString(4, participationStatus.name)
            ps.executeUpdate()
        }
    }

    override fun isKnownDatasource(
        studyId: UUID,
        participantId: String,
        sourceDeviceId: String,
    ): Boolean {
        val hds = storageResolver.getPlatformStorage()

        return hds.connection.use { connection ->
            connection.prepareStatement(COUNT_DEVICE_ID).use { ps ->
                ps.setObject(1, studyId)
                ps.setString(2, participantId)
                ps.setString(3, sourceDeviceId)
                ps.executeQuery().use { rs ->
                    check(rs.next()) { "No count returned for study=$studyId, participant=$participantId, sourceDeviceId=$sourceDeviceId" }
                    ResultSetAdapters.count(rs) > 0 //could also check equal to one, but unique index exists in db
                }
            }
        }
    }

    override fun isKnownParticipant(studyId: UUID, participantId: String): Boolean {
        val hds = storageResolver.getPlatformStorage()

        return hds.connection.use { connection ->
            connection.prepareStatement(COUNT_STUDY_PARTICIPANTS).use { ps ->
                ps.setObject(1, studyId)
                ps.setString(2, participantId)
                ps.executeQuery().use { rs ->
                    check(rs.next()) { "No count returned for study=$studyId, participant=$participantId" }
                    ResultSetAdapters.count(rs) > 0 //could also check equal to one, but unique index exists in db
                }
            }
        }
    }

    override fun getParticipant(studyId: UUID, participantId: String): Participant {
        val hds = storageResolver.getPlatformStorage()

        val (participationStatus, candidateId) = hds.connection.use { connection ->
            connection.prepareStatement(GET_PARTICIPANT).use { ps ->
                ps.setObject(1, studyId)
                ps.setString(2, participantId)
                ps.executeQuery().use { rs ->
                    check(rs.next()) { "No row returned for study=$studyId, participant=$participantId" }
                    ResultSetAdapters.participantStatus(rs) to ResultSetAdapters.candidateId(rs)
                }
            }
        }
        return Participant(participantId, candidateManager.getCandidate(candidateId), participationStatus)
    }

    override fun getParticipationStatus(
        studyId: UUID,
        participantId: String,
    ): ParticipationStatus {
        logger.info(
            "getting participation status" + ChronicleServerUtil.STUDY_PARTICIPANT, studyId,
            participantId
        )

        val hds = storageResolver.getPlatformStorage()

        return hds.connection.use { connection ->
            connection.prepareStatement(GET_PARTICIPATION_STATUS).use { ps ->
                ps.setObject(1, studyId)
                ps.setString(2, participantId)
                ps.executeQuery().use {
                    check(it.next()) { "No row returned for study=$studyId, participant=$participantId" }
                    ResultSetAdapters.participantStatus(it)
                }
            }
        }
    }

    override fun getStudyParticipantIds(studyId: UUID): Set<String> {
        TODO("Not yet implemented")
    }

    override fun getStudyParticipants(studyId: UUID): Set<Participant> {
        TODO("Not yet implemented")
    }

    override fun studyExists(studyId: UUID): Boolean {
        TODO("Not yet implemented")
    }

    override fun getOrganizationIdForStudy(studyId: UUID): UUID {
        TODO("Not yet implemented")
    }


}
